expressions/methods.md

```diff
@@ -29,14 +29,16 @@ After the return value there's a `=>` and then finally the function body. The va
 
 If you want to exit a function early then use the `return` command. If the function has a return type then you need to provide a value to return. If the function does not have a return type then `return` should appear on its own, without a value.
 
-```pony
-class C
-  fun factorial(x: U32): U32 =>
-    if x == 0 then
-	  return 1
-	end
+Pony applies tail call optimization to allow a recursive implementation such as the following factorial function:
 
-	x * factorial(x - 1)
+```pony
+fun factorial(x: I32): I32 ? =>
```

Comment:
I'd say if you want to rely on tail call optimisation, your function should be tail-recursive at the source code level.

A compiler check would be complex to implement. All optimisations of Pony code are made through LLVM and what you're suggesting is non-trivial.

To better convey the idea, I think the code example for this should be a tail-recursive function.